### PR TITLE
eagerly do boundscheck when indexing into CartesianIndices

### LINKED ISSUE #42115: `@inbounds` isn't propagated for `CartesianIndices`
Recent discussion in ArrayInterface.jl demonstrated that `@inbounds` isn't fully propagated when using `CartesianIndices` to convert an integer to a `CartesianIndex`.
Examples across multiple versions and hardware are more fully demonstrated here https://github.com/JuliaArrays/ArrayInterface.jl/pull/199). 

I think the offending line is the missing `@propagate_inbounds` here:

https://github.com/JuliaLang/julia/blob/027071f0fe1a4313ddba5a0909b6f851dff96452/base/abstractarray.jl#L1254

This could probably be fixed here also:

https://github.com/JuliaLang/julia/blob/027071f0fe1a4313ddba5a0909b6f851dff96452/base/abstractarray.jl#L1242

I think @timholy and @mbauman wrote the original code here.
Can we just put `@propagate_inbounds` in front of these?

### PR DESCRIPTION
**Pull request by johnnychen94**
This PR eagerly does boundscheck when doing `getindex(::CartesianIndices, ...)`, this generates simpler inlined codes and furthermore enables SIMD

```julia
julia> versioninfo()
Julia Version 1.8.0-DEV.473
Commit fce2daff16* (2021-09-04 09:37 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin20.6.0)
  CPU: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.1 (ORCJIT, skylake)
```

```julia
function sumcart_oneto(X)
    out = zero(eltype(X))
    R = CartesianIndices(X) # OneTo
    @inbounds @simd for i in eachindex(X)
        out += X[R[i]]
    end
    return out
end

function sumcart_iur(X)
    out = zero(eltype(X))
    R = CartesianIndices(map(i->Base.IdentityUnitRange(1:i), size(X))) # IdentityUnitRange
    @inbounds @simd for i in eachindex(X)
        out += X[R[i]]
    end
    return out
end

function sumcart_unitrange(X)
    out = zero(eltype(X))
    R = CartesianIndices(map(i->1:i, size(X))) # UnitRange
    @inbounds @simd for i in eachindex(X)
        out += X[R[i]]
    end
    return out
end

X = rand(64, 64);

@btime sumcart_oneto($X);
# PR: 241.691 ns (0 allocations: 0 bytes)
# master: 10.526 μs (0 allocations: 0 bytes)
@btime sumcart_iur($X);
# PR: 236.788 ns (0 allocations: 0 bytes)
# master: 10.950 μs (0 allocations: 0 bytes)
@btime sumcart_unitrange($X);
# PR: 234.622 ns (0 allocations: 0 bytes)
# master: 10.529 μs (0 allocations: 0 bytes)
```

The earliest version that I obverse this performance regression is in Julia 1.4.0

A related benchmark:

```julia
function sumcart_linear(X)
    out = zero(eltype(X))
    @inbounds @simd for i in 1:length(X)
        out += X[i]
    end
    return out
end

@btime sumcart_linear($X);
# master: 219.778 ns (0 allocations: 0 bytes)
```

fixes #42115

### CODE REVIEW

File: base/multidimensional.jl [hunk 1]
```diff
@@ -358,8 +358,12 @@ module IteratorsMD
         CartesianIndex()
     end
     @propagate_inbounds function Base.getindex(iter::CartesianIndices{N,R}, I::Vararg{Int, N}) where {N,R}
-        CartesianIndex(getindex.(iter.indices, I))
+        CartesianIndex(_get_cartesianindex.(iter.indices, I))
     end
+    # specialize to gives better hint to compiler to enable SIMD (#42115)
+    @inline _get_cartesianindex(::OneTo, i::Int) = i
+    @inline _get_cartesianindex(::Base.IdentityUnitRange, i::Int) = i
+    @inline _get_cartesianindex(r, i) = getindex(r, i)
```
Comment (johnnychen94) [inline base/multidimensional.jl:366]: A perhaps more appropriate place to make modifications is adding `@propagate_inbounds` in https://github.com/JuliaLang/julia/blob/876df7929fd96de2975bf172f6034ba3e4e1b73e/base/range.jl#L879-L914

Comment (Tokazama) [inline base/multidimensional.jl:366]: I think it would make sense to also have `@inline _get_cartesianindex(::Base.Slice, i::Int) = i`, since this is what slice does when inbounds.

**Review by Tokazama — COMMENTED** [review]
Thanks for doing this!

File: base/multidimensional.jl (hunk 1 above)
Comment (johnnychen94) [reply base/multidimensional.jl:366]: The root issue here is that `throw_boundserror(v, i)` is marked as `@noinline` and thus disables the propagation of `@inbounds`

https://github.com/JuliaLang/julia/blob/8812c5cc242435bf267285fcf3e41e632c84e52c/base/abstractarray.jl#L691

Comment (timholy) [reply base/multidimensional.jl:366]: I don't understand why an error path would do this. But shouldn't the whole expression be in the `@boundscheck`? That is, instead of
```julia
@boundscheck ((i > 0) & ok) || throw_boundserror(v, i) 
```
you just need an extra set of parentheses:
```
@boundscheck(((i > 0) & ok) || throw_boundserror(v, i))
```

Comment (johnnychen94) [reply base/multidimensional.jl]: You're right, this `@boundscheck` does get eliminated if we check individual `getindex(::OneTo, i)`, but it doesn't get eliminated if we check `getindex(::CartesianIndices, I...)`

```julia
getindex_inbounds(r, i...) = @inbounds getindex(r, i...)
R = CartesianIndices((2, 3))
```

```julia
julia> @code_llvm getindex_inbounds(R.indices[1], 1)
;  @ REPL[18]:1 within `getindex_inbounds`
define i64 @julia_getindex_inbounds_2636([1 x i64]* nocapture nonnull readonly align 8 dereferenceable(8) %0, i64 signext %1) #0 {
top:
  ret i64 %1
}

julia> @code_llvm getindex_inbounds(R, 1, 1)
...

L12:                                              ; preds = %top
; │││││└└└
; │││││┌ @ broadcast.jl:642 within `_broadcast_getindex`
; ││││││┌ @ broadcast.jl:666 within `_getindex`
; │││││││┌ @ broadcast.jl:621 within `_broadcast_getindex`
; ││││││││┌ @ tuple.jl:29 within `getindex`
           %9 = getelementptr inbounds [1 x [2 x [1 x i64]]], [1 x [2 x [1 x i64]]]* %1, i64 0, i64 0, i64 0
; │││││└└└└
; │││││┌ @ broadcast.jl:643 within `_broadcast_getindex`
; ││││││┌ @ broadcast.jl:670 within `_broadcast_getindex_evalf`
; │││││││┌ @ range.jl:901 within `getindex`
          %10 = call nonnull {}* @j_throw_boundserror_2682([1 x i64]* nocapture nonnull readonly %9, i64 signext %2) #0
          call void @llvm.trap()
          unreachable
...

L26:                                              ; preds = %L18
          %17 = call nonnull {}* @j_throw_boundserror_2681([1 x i64]* nocapture readonly %11, i64 signext %3) #0
          call void @llvm.trap()
          unreachable
...
```

Notice the two `throw_boundserror` call. Thus I did an eager check so that we can pass `@inbounds` hint through the broadcasting/map. After 20c4be5:

```julia
julia> @code_llvm getindex_inbounds(R, 1, 1)
;  @ REPL[18]:1 within `getindex_inbounds`
define void @julia_getindex_inbounds_2685([1 x [2 x i64]]* noalias nocapture sret([1 x [2 x i64]]) %0, [1 x [2 x [1 x i64]]]* nocapture nonnull readonly align 8 dereferenceable(16) %1, i64 signext %2, i64 signext %3) #0 {
top:
  %.sroa.0.sroa.0.0..sroa.0.0..sroa_cast4.sroa_idx = getelementptr inbounds [1 x [2 x i64]], [1 x [2 x i64]]* %0, i64 0, i64 0, i64 0
  store i64 %2, i64* %.sroa.0.sroa.0.0..sroa.0.0..sroa_cast4.sroa_idx, align 8
  %.sroa.0.sroa.2.0..sroa.0.0..sroa_cast4.sroa_idx8 = getelementptr inbounds [1 x [2 x i64]], [1 x [2 x i64]]* %0, i64 0, i64 0, i64 1
  store i64 %3, i64* %.sroa.0.sroa.2.0..sroa.0.0..sroa_cast4.sroa_idx8, align 8
  ret void
}
```

**Review by timholy — COMMENTED** [review]
LGTM. Can you add a test in `test/boundscheck_exec.jl`?

File: base/multidimensional.jl [hunk 2]
```diff
@@ -357,21 +357,32 @@ module IteratorsMD
...
+            @inbounds getindex(r, i)
+        end
+        CartesianIndices(indices)
     end
     @propagate_inbounds function Base.getindex(iter::CartesianIndices{N},
         C::CartesianIndices{N}) where {N}
-        CartesianIndices(getindex.(iter.indices, C.indices))
+        getindex(iter, C.indices...)
     end
+    @inline Base.getindex(iter::CartesianIndices{0}, ::CartesianIndices{0}) = iter
```
Comment (johnnychen94) [inline base/multidimensional.jl:385]: If we want to backport to 1.6 or 1.7, then these changes should be separated to another PR.

Comment (timholy) [reply base/multidimensional.jl:385]: What's the justification here?
```julia
julia> ()[CartesianIndex()]
ERROR: MethodError: no method matching getindex(::Tuple{}, ::CartesianIndex{0})
Closest candidates are:
  getindex(::Tuple, ::Int64) at tuple.jl:29
  getindex(::Tuple, ::Real) at tuple.jl:30
  getindex(::Tuple, ::Colon) at tuple.jl:33
  ...
Stacktrace:
 [1] top-level scope
   @ REPL[12]:1
```

Comment (johnnychen94) [reply base/multidimensional.jl:385]: This extra line is for the `@test c[c] === c` test case, because `getindex(iter, C.indices...)` will call the `getindex(::CartesianIndices{0})` method, which returns `CartesianIndex()` instead of `[CartesianIndex()]`

https://github.com/JuliaLang/julia/blob/8812c5cc242435bf267285fcf3e41e632c84e52c/test/cartesian.jl#L150-L156

**Comment by johnnychen94** [comment]
The `getindex(::CartesianIndices, ::RangeTypes)` methods are newly introduced in https://github.com/JuliaLang/julia/pull/40344 so I've removed them in d58ce6a. Doing this we can backport this PR to 1.6 and 1.7 branch. I'll submit a separate PR after this.

**Comment by johnnychen94** [comment]
I think this one is ready.

cc: @timholy @mbauman

**Comment by timholy** [comment]
Thanks!

**Comment by mbauman** [comment]
Aha, this makes sense! Thank you!
